### DISCUSSION TITLE
Support pypy binary installs without python-build

### DISCUSSION THREAD

**Discussion by scop** [jdx/mise#3866] [Ideas]
Would be nice to support pypy installs without delegating to python-build, i.e. with `python.compile` set to false in settings.

So e.g. `mise install python@pypy3.10-v7.3.17` would work.

- release tags: https://github.com/pypy/pypy/tags
- binary downloads: https://pypy.org/download.html

**Comment by JamBalaya56562** [comment]
This is implemented now — #11846, merged as `7b3514436dad`. It is not in a release yet: `v2026.8.4` predates the merge, so it will land in the next one.

`python.compile = false` installs PyPy straight from `downloads.python.org/pypy`, with python-build out of the picture:

```
# before
$ MISE_PYTHON_COMPILE=false mise install python@pypy3.10-7.3.17
mise ERROR no precompiled python found for core:python@pypy3.10-7.3.17
```

```
# after
$ mise install python@pypy3.10-7.3.17
download pypy3.10-v7.3.17-win64.zip
generate checksum / extract / ensurepip
python --version
  Python 3.10.14 (39dc8d3c85a7, Aug 27 2024, 14:33:33)
  [PyPy 7.3.17 with MSC v.1929 64 bit (AMD64)]

$ mise exec python@pypy3.10-7.3.17 -- pip --version
pip 23.0.1 from …\installs\python\pypy3.10-7.3.17\lib\site-packages\pip (python 3.10)
```

**What I actually ran, and what I did not.** That transcript is from a Windows host, which is the platform this unblocks most (see below). The `MISE_PYTHON_COMPILE=false mise install …` invocation on Linux is served by the same code path, but I did not run it end to end — no Linux host. What I did measure on Linux is the part that differs per platform: the archive layout (`pypy3.10-v7.3.17-linux64.tar.bz2` unpacks to one top-level dir, and `bin/` already carries the `python → pypy3.10` symlinks) and `./bin/python -m ensurepip --upgrade --default-pip` succeeding offline from the bundled wheel.

**One detail against your example:** the string mise lists is `pypy3.10-7.3.17`, **without the `v`** — the spelling python-build uses, so the same string resolves on either path. `mise ls-remote python` is the authority.

**Windows gets more out of this than the setting implies.** The precompiled path is taken *unconditionally* there, so before this change pypy versions never appeared in `mise ls-remote python` on Windows at all and could not be installed by any route. Against the live index: 87 pypy entries on `win64/x64`, 94 on `linux/x64`.

`python@latest` is unaffected — the two halves are concatenated with CPython last, so the tail is always CPython.

Limits, so the scope is clear:

- The **default** (compiled) path on unix still goes through python-build for pypy. This change only touches the precompiled path, so installs that work today are untouched. Serving that natively too would drop the pyenv clone for pypy users as well — a reasonable follow-up.
- PyPy publishes no machine-readable checksums (the sidecar files 404, `versions.json` has no hash field), so this uses mise's standard treatment for such sources: a blake3 digest recorded in the lockfile on first install and enforced on every install after, the same as `ubi:`. Parsing `pypy.org/checksums.html` is a strictly additive follow-up.
- Nightlies are excluded — they are rolling artifacts, so a pinned name would point at bytes that change daily.
- No provenance step; PyPy publishes no attestations.
- `pypy3.6` has no `win64` build, so it is simply not listed on Windows rather than resolving to something that cannot install. musl hosts are unchanged, since PyPy publishes glibc builds only.
